---
title: "Differential Gene Expression Lab 3"
author: "Sharon"
date: "2025-02-20"
output: html_document
---

This markdown is to analyse sequencing result.

This code was written by me while following the instructions provided by the lecturer for Genomics at ANU. The methodology and structure were outlined in the course, but all implementation was done by me.


```{r, echo=FALSE}
library(tidyverse)
library(edgeR)
library(RColorBrewer)
library(limma)
```


```{r}
designFile <- read.csv("Sample Sheet lab3 RNAseq - Sheet1.csv", nrow=73)

# remove samples without a treatment
dim(designFile)[1] # number samples
designFile = designFile[designFile$treatment != "",]
designFile = designFile[designFile$treatment != " ",]
dim(designFile)[1] # number samples

# get list of all gene count files
# some samples wont have counts, so need to do intersection
tmp1 <- list.files("gene-counts/")
tmp2 = designFile$filename
countFiles = intersect(tmp1,tmp2)

## reading in all files
dge <- readDGE(files = countFiles, path = "gene-counts", columns = c(1,2),sep=' ') 

# looks for bad samples
hist(dge$samples$lib.size,breaks=20)

## cut at 1000, try other thresholds
# check how many sample remain at different levels of filtering?
thresholds=1000 # can be changed
abline(v=thresholds)
table(dge$samples$lib.size > thresholds)
```

```{r}
# now read in only the good files
dge <- readDGE(files = countFiles[dge$samples$lib.size > thresholds], path = "gene-counts", columns = c(1,2), sep=' ')
dim(dge)


# this is key to setting up the sample contrasts, linking sample names with treatment
dge$samples$sampleGroup <- designFile$treatment[match(dge$samples$files, designFile$filename)]
dge$samples$sampleGroup


# experimental design matrix
treat <- dge$samples$sampleGroup
design <- model.matrix(~0 + treat)
design
```

```{r}
#Read counts per million, try higher or lower counts and samples
## Abundance filter (CPM > 3 in at least 6 samples)
keep <- rowSums(cpm(dge) > 3) > 6
table(keep)

# cutting down from 14466 genes
dge <- dge[keep, ]


## Re-calculate lib size based on retained transcripts
dge$samples$lib.size <- colSums(dge$counts)


## TMM Normalization (normalize by library size)
dge <- calcNormFactors(dge, method = "TMM")


## NB-dispersion
dge <- estimateDisp(dge, design, robust=TRUE)


## define colour per sample group
col1 <- c("#FF0000","#0000FF")
col1 <- col1[factor(dge$samples$sampleGroup)]


## Multi-Dimensional Scaling plots
# check how the samples cluster by treatment or species
# MDS is dimension reduction from all genes to 2 dimensions, then plotted in a sample space
mds <- plotMDS(dge, dim.plot = c(1,2), col=col1, labels = dge$samples$treat)

```

```{r}
### fit NB-GLM
fit <- glmQLFit(dge, design, robust=TRUE)

## plot NB-GLM variance
plotQLDisp(fit) # check how the variance is distributed across low and high expression genes
summary(fit$df.prior)


```

```{r}
## setup contrasts for differential expression testing
my.contrasts <- makeContrasts(
  treatment = treatA - treatB,
  levels=design
)
### perform GLMs for all contrasts specified and collate DEGs
for(i in colnames(my.contrasts)){
  res <- glmQLFTest(fit, contrast = my.contrasts[,i])
  # these are the differentially expressed genes, Up in heat treatment.
  plotMD(res, main = i, ylim = c(-12, 12))
  abline(h = c(-1, 1), col = c("goldenrod", "goldenrod"), lty=2)
}


```

```{r}
# write out the results table with gene names, then you can sort by P value to find the differential genes
out.file <- cbind(rownames(res$table),res$table)
write_csv(out.file, file="3161RNAlong_output.csv")

# Notice anything expected in the gene names??
tmp = out.file[out.file$PValue <= 0.05,]
tmp[tmp$logFC <= -2,]
tmp[tmp$logFC >= 2,]

head(tmp[order(tmp$logFC),],10)
tail(tmp[order(tmp$logFC),],10)
## eg. LOC663293
```

